### PR TITLE
[HCP Telemetry] Move first TelemetryConfig Fetch into the TelemetryConfigProvider

### PR DESCRIPTION
**Pull request by Achooo**
### Description

This is a follow up refactor for PR  #18168 

Now that we have the TelemetryConfigProvider that continuously fetches config from HCP, we can skip the `deps.go` first fetch and leave that work to the Provider.

Changes:
1. The provider will make a first fetch during initialization to see if there is config available.
- If none, it will set defaults and run the provider
2. OTELSink and OTELExporter : Add an interface to disable, as the sink can now be turned on/off based on the config. Metrics will not be recorded if the sink is not enabled.

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

### CODE REVIEW

File: agent/hcp/telemetry/otel_sink.go
Comment (Achooo) [inline agent/hcp/telemetry/otel_sink.go:25]: @chapmanc Although I like this idea a lot, I think it might be neater to use a Disabler interface 😓 I know it's less intuitive in the telemetry_config_provider, but it's leading to a lot more complexity in the tests etc of the sink.

I think as functionality, to forcefully disable makes more sense to me starting off when using the sink, our use case is just a bit different here.

File: agent/hcp/telemetry/otel_sink.go
Comment (Achooo) [inline agent/hcp/telemetry/otel_sink.go:106]: Moved this in here to simplify the deps. We could also remove it fully.

File: agent/hcp/deps.go
Comment (Achooo) [inline agent/hcp/deps.go:1]: The entire first fetchTelemetryConfig call is removed simplifying this file quite a lot.

File: agent/hcp/telemetry/otel_sink.go
Comment (Achooo) [inline agent/hcp/telemetry/otel_sink.go:170]: How the early return works when it is not enabled

File: agent/hcp/telemetry/otel_exporter.go
Comment (Achooo) [inline agent/hcp/telemetry/otel_exporter.go:69]: Early return when not enabled.

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:24]: @jjti had this filter as a default in his example PR, I'm not sure we want to collect metrics at the start?

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:77]: Provide sane defaults to avoid nil pointers in the Get* methods - cleaner than nil checks everywhere.

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:87]: I put the first fetch in the New initialization so that we can test the `run` function more easily.

File: agent/hcp/deps.go
Comment (jjti) [inline agent/hcp/deps.go:94]: nit: this is the only thing here that could throw an error, so this def a nit, but I slightly prefer what you had before which was more explicit re: catching the `NewOTELSink` error and returning a wrapped version with `fmt.Errorf("failed to create OTELSink: %w", err)`

eg of justification about always wrapping: https://github.com/tomarrell/wrapcheck
Uber's style guide that's more case dependent (only recommending if the enclosed error doesn't have enough info by itself): https://github.com/uber-go/guide/blob/master/style.md#error-wrapping

File: agent/hcp/telemetry_provider.go
Comment (jjti) [inline agent/hcp/telemetry_provider.go:70]: I think this is redundant since it's set in `getUpdate`:

https://github.com/hashicorp/consul/blob/9ed67e439632f3e29f0b8c1fcf548be7f3588b9d/agent/hcp/telemetry_provider.go#L129

relatedly, I think `getUpdate` could be renamed to something like `updateConfig` since it's mutating config internally

**Review by jjti — APPROVED** [review]
LGTM. One minor comment

File: agent/hcp/telemetry/otlp_transform.go
Comment (Achooo) [inline agent/hcp/telemetry/otlp_transform.go:16]: Just a small cleanup to follow convention do `err +<name>`

File: agent/hcp/client/telemetry_config.go
Comment (Achooo) [inline agent/hcp/client/telemetry_config.go:20]: Export this to use as default in https://github.com/hashicorp/consul/pull/18318/files#diff-9fd2535b13a1b9701661bdd379418cb004a41b5b747ff9911e8cbbffea54b9afR60.

In case the first fetch is not available, this default is set to all metrics.

File: agent/hcp/telemetry/otel_exporter.go
Comment (Achooo) [inline agent/hcp/telemetry/otel_exporter.go:1]: `EndpointProvider` now embeds the `Enabler` to turn on/off export of metrics.

File: agent/hcp/deps.go
Comment (Achooo) [reply agent/hcp/deps.go]: Thanks! Fixed 💯

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [reply agent/hcp/telemetry_provider.go]: 100%, I also renamed the method, thanks for catching this

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:70]: Move this within the run

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:84]: just return the interval directly

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:94]: return interval

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:47]: Cleanup: privatize these fields after removing hashing

File: agent/hcp/telemetry/otlp_transform.go
Comment (jjti) [inline agent/hcp/telemetry/otlp_transform.go:99]: nit, also applying to other errs below

```suggestion
			return out, fmt.Errorf("failed to convert metric to otel format: %w: %T", errTemporality, a)
```

**Review by jjti — APPROVED** [review]
LGTM. Notes on how this was tested appreciated tho!

File: agent/hcp/client/telemetry_config.go
Comment (Achooo) [inline agent/hcp/client/telemetry_config.go:51]: We could also get rid of this function in entirely and let the `TelemetryConfigProvider` check the endpoint for `nil` directly?

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:47]: Lowercased these as we don't do the hashing anymore. Missed this in the previous PR.

File: agent/hcp/telemetry/otel_sink.go
Comment (Achooo) [inline agent/hcp/telemetry/otel_sink.go:24]: Went back and forth between "`Disabler`, `Disablable` which would allow is to have just an `Enabled()` method...

This one makes the method `IsEnabled()` but the name is better?

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:84]: Missed this in previous PR, but we can really just return the refreshInterval to reset the ticker instead of the whole config.

File: agent/hcp/client/telemetry_config.go
Comment (chapmanc) [inline agent/hcp/client/telemetry_config.go:52]: I'll be honest that I am a bit confused as to why the MetricsConfig doesn't ahve a disabled field on it? It has everything else from the agent telemetry config so it seems like we should just also add disabled to it as well? Maybe you can validate my thinking there?

File: agent/hcp/telemetry/otel_exporter.go
Comment (chapmanc) [inline agent/hcp/telemetry/otel_exporter.go:22]:
```suggestion
// This abstraction layer offers flexibility, in particular for dynamic configuration or changes to the endpoint. The OTEL exporter will call the Disabled interface to validate that it should actually export metrics. 
```

File: agent/hcp/telemetry/otlp_transform.go
Comment (chapmanc) [inline agent/hcp/telemetry/otlp_transform.go:110]: Nit: two colons

File: agent/hcp/telemetry/otlp_transform.go
Comment (chapmanc) [inline agent/hcp/telemetry/otlp_transform.go:119]: Nit: two colons

**Review by chapmanc — APPROVED** [review]
Small nits but looks good to me.

File: agent/hcp/client/telemetry_config.go
Comment (Achooo) [reply agent/hcp/client/telemetry_config.go:52]: ah oops yep no this is just me being a dummy, I will just add it there!

Comment (chapmanc) [reply agent/hcp/client/telemetry_config.go:52]: Okay and then that endpoint check will be the disabled flag check?

Comment (Achooo) [reply agent/hcp/client/telemetry_config.go]: I've just updated the PR!

File: agent/hcp/telemetry/otlp_transform.go
Comment (Achooo) [reply agent/hcp/telemetry/otlp_transform.go]: thanks good catch!

File: agent/hcp/client/telemetry_config.go
Comment (Achooo) [inline agent/hcp/client/telemetry_config.go:85]: I feel like we should just set this field TGW side in the response for :
A) Cleaner code in Consul
B) Better control (feels weird that its dependent on the endpoint)
Will discuss and get back to this comment, shouldn't be a big code change, just setting `Disabled` directly from the payload response. But this requires changes TGW side.

**Review by loshz — APPROVED** [review]

**Comment by Achooo** [comment]
Will hold off on merging this for now - and won't be in the Tuesday Consul release.

**Why?** In order to test E2E more thoroughly

- Confirmed that we can do an `enabled` field TGW side. 
- Will also update this PR  with that. This will allow me to E2E test easily by turning on/off metrics TGW side.

**Comment by Achooo** [comment]
Changes to PR:
1. Merge `main` changes, commits:
-  [9c8711c](https://github.com/hashicorp/consul/pull/18318/commits/9c8711c5911fee08d8d5fc4c57d5db91f804b306) and - [150fb64](https://github.com/hashicorp/consul/pull/18318/commits/150fb649bdb2d9fa4dedcf6b355c3be61fc04011)
2. Update hcp-sdk-go to new version `[v0.61.0](https://github.com/hashicorp/consul/pull/18318/commits/1b3d71a20ed46952afdb809488b846623bfb77a2)` to bring in the SDK changes from CTGW updates:
- [1b3d71a](https://github.com/hashicorp/consul/pull/18318/commits/1b3d71a20ed46952afdb809488b846623bfb77a2)
3. Use `disabled flag` to remove endpoint empty checks:
- [4769d7b](https://github.com/hashicorp/consul/pull/18318/commits/4769d7b3ddf692f70138295cd3491a88aaa24a86)
4. Update Telemetry Provider to handle an edge case 401 and 404, when someone unlinks a cluster from HCP. We want to disable metrics in this case.
- [89cb0d3](https://github.com/hashicorp/consul/pull/18318/commits/89cb0d3248eaacc898bb720607dbd61d536287b3)

I will add end to end screenshots in a next comment!

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:116]: @jjti I emit a success metric in this case, since it's not really a failure?

Would it be better to emit a different custom metric, e.g. `internalMetricRefreshDisabled []string = []string{"hcp", "telemetry_config_provider", "refresh", "disabled"}`

My reasoning is that we should treat it as a succesful response and makes it easier to calculate success rate. Otherwise we have to know about this edge case and consider it.

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [inline agent/hcp/telemetry_provider.go:114]: @jjti only targeting 404 and 401, but lmk if you want us to target 4XX class more generally? I don't think we need to care about the other ones but you might have more CCM context than me!

File: agent/hcp/client/telemetry_config.go
Comment (Achooo) [inline agent/hcp/client/telemetry_config.go:53]: No need for this MetricsEnabeld func and the empty endpoint logic since we now have a `disabled` flag

**Review by jjti — COMMENTED** [review]
Few small things but mostly LGTM

File: agent/hcp/telemetry_provider.go
Comment (jjti) [reply agent/hcp/telemetry_provider.go:114]: > if you want us to target 4XX class more generally?

I'm open to other thoughts, but imo we should disable on any client-side status code response (4XX). All of them suggest the client was unable to get the agent telemetry config for some reason related to this configuration request, rather than a server side issue

File: agent/hcp/telemetry_provider.go
Comment (jjti) [reply agent/hcp/telemetry_provider.go:116]: It's up to you here, either work. One other option is to use `metrics.IncrCounterWithLabels([]string{"hcp", "telemetry_config_provider", "refresh"}, 1, []metrics.Label{Name:"result", Value:"skipped"})`, etc. So put the result in a label so there's just one counter, but multiple active time series, for the refreshing

You can see a bit of "use the labels to delineate result" here in this (beta) metric in consul that's emitted from all rpc endpoints: https://developer.hashicorp.com/consul/docs/agent/telemetry#server-workload

I prefer it personally to the rest of the Consul metrics that are like `consul.cache.fetch_success` vs `consul.cache.fetch_error` etc

File: agent/hcp/client/telemetry_config.go
Comment (jjti) [reply agent/hcp/client/telemetry_config.go:97]: @Achooo pointed out these aren't in the cloud-sdk clients. Which is lame and weird and less safe than the clients we generated internally. Given more time down the line I'll check out if we can fix cloud-sdk to include the nil-safe Getters, but they're definitely not here rn

Comment (Achooo) [reply agent/hcp/client/telemetry_config.go:97]: f you ran this test continuously at some point you would hit a flaky test without that mock statement. We also don't want to assert on it, because it won't necessarily be called.

More context:
1. In deps.go, `NewHCPProvider` is called https://github.com/hashicorp/consul/pull/18318/files#diff-d652bfc7ebdcf8edacd32ad13fd9059b666981a2a76ef9bc5d985be7edac5601R68
2. In TelemetryConfigProvider - a goroutine starts to fetch TelemetryConfig from HCP: https://github.com/hashicorp/consul/pull/18318/files#diff-9fd2535b13a1b9701661bdd379418cb004a41b5b747ff9911e8cbbffea54b9afR78
3. This run function might make a call using the client:
https://github.com/hashicorp/consul/pull/18318/files#diff-9fd2535b13a1b9701661bdd379418cb004a41b5b747ff9911e8cbbffea54b9afR86

If (3) happens by the time the test ends, we need to mock that client call. It could also be that the test passes before getting there - no call is ever made. 

Maybe I should clarify this with a comment.

File: agent/hcp/deps_test.go
Comment (Achooo) [reply agent/hcp/deps_test.go:26]: If you ran this test continuously at some point you would hit a flaky test without that mock statement. We also don't want to assert on it, because it won't necessarily be occur.

More context:
1. In deps.go, `NewHCPProvider` is called https://github.com/hashicorp/consul/pull/18318/files#diff-d652bfc7ebdcf8edacd32ad13fd9059b666981a2a76ef9bc5d985be7edac5601R68
2. In TelemetryConfigProvider's `NewHCPProvider` a goroutine starts to fetch TelemetryConfig from HCP: https://github.com/hashicorp/consul/pull/18318/files#diff-9fd2535b13a1b9701661bdd379418cb004a41b5b747ff9911e8cbbffea54b9afR78
3. This run function might make a call using the client:
https://github.com/hashicorp/consul/pull/18318/files#diff-9fd2535b13a1b9701661bdd379418cb004a41b5b747ff9911e8cbbffea54b9afR86

If (3) happens by the time the test ends, we need to mock that client call. It could also be that the test passes before getting there - no call is ever made. 

Maybe I should clarify this with a comment.

Comment (Achooo) [reply agent/hcp/deps_test.go:26]: Alternatively, we could maybe move this NewHCPProvider  call to the upper level NewDeps and pass it in as a param to the sink function. We can then just pass in a mock provider as a param to completely avoid this issue, as it might be confusing for someone reading the code.

I think that's likely better, what do you think?

p.s. a note that the NewDeps function has no test since a while ago... but if it was to be tested, we would face the same issue.

File: agent/hcp/client/telemetry_config.go
Comment (Achooo) [reply agent/hcp/client/telemetry_config.go:97]: Whoops wrong comment, moved the above to the right comment

**Comment by Achooo** [comment]
## End to End Tests

### Scenario 1: Disable and Re-enable metrics
1. Start a self-managed cluster
2. Disable metrics using LD flag (`disabled: true`)
3. Re-enable metrics using LD flag (`disabled: false`) after a few minutes
Screenshot 2023-08-29 at 12 43 21 PM

🥳 You can see the dip (without metrics) when I disabled then re-enabled metrics in HCP Observability.

### Scenario 2: Regression testing - addtional metric filters (dynamically change metric filters)
1. Add a new metric filter `consul.state`
2. After some time, add an additional metric filter `consul.acl`

Screenshot 2023-08-29 at 12 54 55 PM

🥳 You can see the new metrics on the right pop up staggered, showing that they were dynamically toggled

### Scenario 3: Unlink cluster
1. Proceed to unlink this cluster in HCP
Screenshot 2023-08-29 at 12 56 43 PM
Screenshot 2023-08-29 at 1 28 13 PM

🥳 You can see the metrics in Grafana stop after unlinking.

File: agent/hcp/client/telemetry_config.go
Comment (chapmanc) [inline agent/hcp/client/telemetry_config.go:116]: Like known errors. Good stuff.

File: agent/hcp/deps_test.go
Comment (chapmanc) [reply agent/hcp/deps_test.go:26]: I agree. I would do the final suggestion mentioned.

File: agent/hcp/telemetry_provider.go
Comment (chapmanc) [reply agent/hcp/telemetry_provider.go:114]: I'd vote for any 4xx codes.

File: agent/hcp/telemetry_provider.go
Comment (chapmanc) [reply agent/hcp/telemetry_provider.go:116]: I don't feel strongly either way but I feel like a label is a bit odd. I think if we want to explicitly measure the case of 4xx vs 5xx vs legitmately successful updates we could do it with different metrics. Wouldn't die on a hill and will leave it up to the observability team.

**Review by chapmanc — APPROVED** [review]
LGTM some small changes.

File: agent/hcp/telemetry_provider.go
Comment (jjti) [reply agent/hcp/telemetry_provider.go:116]: I also wouldn't die on this hill. I haven't seen a recommendation for one approach vs other (they result in the same total sample count + cost in a DD or AMP). I slightly prefer the label tho. Eg from an Envoy HTTP response counter:

Screenshot 2023-08-29 at 1 58 01 PM

File: agent/hcp/telemetry/otel_exporter.go
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go:73]: Let's add it in a follow-up if a need arises?

Comment (jjti) [reply agent/hcp/telemetry/otel_exporter.go:73]: I was thinking as a counterpart to the existing metrics we emit here: `goMetrics.IncrCounter(internalMetricExportFailure, 1)` + `goMetrics.IncrCounter(internalMetricExportSuccess, 1)`

File: agent/hcp/telemetry_provider.go
Comment (Achooo) [reply agent/hcp/telemetry_provider.go:116]: After walking through an example, there are reasons why this is a good idea. I will target a follow-up PR to do this since we need to modify all other metrics to follow the same pattern (for consistency_

File: agent/hcp/telemetry/otel_exporter.go
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go:73]: Sounds good 👍🏽 will address in follow-up PR for metric key + labels.

**Review by loshz — APPROVED** [review]

**Comment by hc-github-team-consul-core** [comment]
Backport failed @Achooo. Run: https://github.com/hashicorp/consul/actions/runs/6028096530

**Comment by hc-github-team-consul-core** [comment]
Backport failed @Achooo. Run: https://github.com/hashicorp/consul/actions/runs/6028096530

**Pull request merged**
